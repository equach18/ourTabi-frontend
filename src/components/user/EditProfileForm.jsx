import { useState } from "react";
import { useNavigate } from "react-router-dom";

function EditProfileForm({ updateUser, currentUser }) {
  const navigate = useNavigate();

  const [formData, setFormData] = useState({
    firstName: currentUser.firstName || "",
    lastName: currentUser.lastName || "",
    email: currentUser.email || "",
    profilePic: currentUser.profilePic || "",
    bio: currentUser.bio || "",
  });

  const [formErrors, setFormErrors] = useState([]);
  const [isSubmitting, setIsSubmitting] = useState(false);

  function handleChange(evt) {
    const { name, value } = evt.target;
    setFormData((data) => ({
      ...data,
      [name]: value,
    }));
  }

  async function handleSubmit(evt) {
    evt.preventDefault();
    setFormErrors([]);
    setIsSubmitting(true);

    // Filter unchanged fields
    const updatedFields = {};
    for (let key in formData) {
      if (formData[key] !== currentUser[key]) {
        updatedFields[key] = formData[key];
      }
    }

    if (Object.keys(updatedFields).length === 0) {
      setFormErrors(["No changes made."]);
      setIsSubmitting(false);
      return;
    }

    try {
      const result = await updateUser(updatedFields);
      if (result.success) {
        navigate("/dashboard");
      } else {
        setFormErrors([result.error]);
      }
    } catch (err) {
      setFormErrors([err.message || "Failed to update profile."]);
    } finally {
      setIsSubmitting(false);
    }
  }

  return (
    <div className="h-[calc(100vh-4rem)] flex items-center justify-center px-4 py-10">
      <div className="w-full max-w-xl bg-white p-8 rounded-xl shadow-lg">
        <h1 className="text-3xl font-bold text-gray-600 mb-6">Edit Profile</h1>

        <form onSubmit={handleSubmit} className="space-y-4">
          {/* First Name */}
          <input
            type="text"
            name="firstName"
            placeholder="First Name"
            value={formData.firstName}
            onChange={handleChange}
            className="w-full py-2 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
          />

          {/* Last Name */}
          <input
            type="text"
            name="lastName"
            placeholder="Last Name"
            value={formData.lastName}
            onChange={handleChange}
            className="w-full py-2 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
          />

          {/* Email */}
          <input
            type="email"
            name="email"
            placeholder="Email"
            value={formData.email}
            onChange={handleChange}
            className="w-full py-2 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
          />

          {/* Profile Pic */}
          <input
            type="url"
            name="profilePic"
            placeholder="Profile Picture URL"
            value={formData.profilePic}
            onChange={handleChange}
            className="w-full py-2 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
          />

          {/* Bio */}
          <textarea
            name="bio"
            placeholder="Tell us about yourself..."
            value={formData.bio}
            onChange={handleChange}
            rows="4"
            className="w-full py-2 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
          />

          {/* Errors */}
          {formErrors.length > 0 && (
            <div className="text-red-600 text-sm space-y-1">
              {formErrors.map((err, i) => (
                <p key={i}>{err}</p>
              ))}
            </div>
          )}

          {/* Buttons */}
          <div className="flex justify-end space-x-4 pt-2">
            <button
              type="button"
              onClick={() => navigate("/dashboard")}
              className="bg-zinc-400 text-white px-5 py-2 rounded-lg hover:bg-zinc-500 transition"
            >
              Cancel
            </button>
            <button
              type="submit"
              disabled={isSubmitting}
              className={`bg-blue-500 text-white px-6 py-2 rounded-lg hover:bg-blue-600 transition ${
                isSubmitting && "opacity-50 cursor-not-allowed"
              }`}
            >
              {isSubmitting ? "Saving..." : "Save Changes"}
            </button>
          </div>
        </form>
      </div>
    </div>
  );
}

export default EditProfileForm;
