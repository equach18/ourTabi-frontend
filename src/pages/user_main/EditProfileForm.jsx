import { useState } from "react";
import { useNavigate } from "react-router-dom";

function EditProfileForm({ updateUser, currentUser }) {
  const navigate = useNavigate();

  const [formData, setFormData] = useState({
    firstName: currentUser.firstName || "",
    lastName: currentUser.lastName || "",
    email: currentUser.email || "",
    profilePic: currentUser.profilePic || "",
    bio: currentUser.bio || "",
  });

  const [formErrors, setFormErrors] = useState([]);
  const [isSubmitting, setIsSubmitting] = useState(false);

  function handleChange(evt) {
    const { name, value } = evt.target;
    setFormData((data) => ({
      ...data,
      [name]: value,
    }));
  }

  async function handleSubmit(evt) {
    evt.preventDefault();
    setFormErrors([]);
    setIsSubmitting(true);

    // Filter unchanged fields
    const updatedFields = {};
    for (let key in formData) {
      if (formData[key] !== currentUser[key]) {
        updatedFields[key] = formData[key];
      }
    }

    if (Object.keys(updatedFields).length === 0) {
      setFormErrors(["No changes made."]);
      setIsSubmitting(false);
      return;
    }

    try {
      const result = await updateUser(updatedFields);
      if (result.success) {
        navigate("/dashboard");
      } else {
        setFormErrors([result.error]);
      }
    } catch (err) {
      setFormErrors([err.message || "Failed to update profile."]);
    } finally {
      setIsSubmitting(false);
    }
  }

  return (
    <div className="max-w-lg mx-auto mt-10 p-6 bg-white shadow-lg rounded-lg">
      <h1 className="text-2xl font-bold text-blue-500 mb-4">Edit Profile</h1>

      <form onSubmit={handleSubmit} className="space-y-4">
        {/* First Name */}
        <input
          type="text"
          name="firstName"
          placeholder="First Name"
          value={formData.firstName}
          onChange={handleChange}
          className="w-full p-2 border rounded-lg"
        />

        {/* Last Name */}
        <input
          type="text"
          name="lastName"
          placeholder="Last Name"
          value={formData.lastName}
          onChange={handleChange}
          className="w-full p-2 border rounded-lg"
        />

        {/* Email */}
        <input
          type="email"
          name="email"
          placeholder="Email"
          value={formData.email}
          onChange={handleChange}
          className="w-full p-2 border rounded-lg"
        />

        {/* Profile Picture */}
        <input
          type="url"
          name="profilePic"
          placeholder="Profile Picture URL"
          value={formData.profilePic}
          onChange={handleChange}
          className="w-full p-2 border rounded-lg"
        />

        {/* Bio */}
        <textarea
          name="bio"
          placeholder="Tell us about yourself..."
          value={formData.bio}
          onChange={handleChange}
          className="w-full p-2 border rounded-lg"
          rows="4"
        />

        {/* Errors */}
        {formErrors.length > 0 && (
          <div className="text-red-500">
            {formErrors.map((err, i) => (
              <p key={i}>{err}</p>
            ))}
          </div>
        )}

        {/* Buttons */}
        <div className="flex space-x-4">
          <button
            type="submit"
            disabled={isSubmitting}
            className="bg-blue-500 text-white px-6 py-2 rounded-lg"
          >
            {isSubmitting ? "Saving..." : "Save Changes"}
          </button>
          <button
            type="button"
            onClick={() => navigate("/dashboard")}
            className="bg-gray-400 text-white px-6 py-2 rounded-lg"
          >
            Cancel
          </button>
        </div>
      </form>
    </div>
  );
}

export default EditProfileForm;