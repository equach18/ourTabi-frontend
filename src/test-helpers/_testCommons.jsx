import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import UserContext from "../context/UserContext";
import TripContext from "../context/TripContext";

export const testUser1 = {
  id: 1,
  firstName: "Test1",
  lastName: "User1",
  username: "testUser1",
  profilePic: "",
};
export const testUser2 = {
  id: 2,
  firstName: "Test2",
  lastName: "User2",
  username: "testUser2",
  profilePic: "testUser2.png",
};
export const testUser3 = {
  id: 3,
  firstName: "Test2",
  lastName: "User2",
  username: "testUser3",
  profilePic: "testUser3.png",
};

export const defaultMembers = [
  {
    userId: testUser1.id,
    username: testUser1.username,
    profilePic: testUser1.profilePic,
  },
  {
    userId: testUser2.id,
    username: testUser2.username,
    profilePic: testUser2.profilePic,
  },
];

export const testActivity = {
  id: 1,
  name: "Kayaking",
  tripId: 53,
  category: "Other",
  location: "Chicago",
  description: "So much fun",
  scheduledTime: "2025-07-10T14:30:00Z",
};

export const testComment = {
  id: 1,
  text: "Test comment",
  createdAt: "2025-07-11T14:30:00Z",
  userId: 1,
};

export const testTrip = {
  id: 42,
  title: "Test trip",
  destination: "Chicago, IL",
  startDate: "2025-07-10",
  endDate: "2025-07-12",
  isPrivate: true,
};

export const testFriendsData = {
  friends: [testUser1 ],
  incomingRequests: [testUser2],
  sentRequests: [testUser3],
};

// Default mock context values
const defaultUserContext = { currentUser: testUser1 };
const defaultTripContext = {
  voteOnActivity: vi.fn(),
  votes: {},
  deleteComment: vi.fn(),
  addMember: vi.fn(),
  removeMember: vi.fn(),
  members: defaultMembers,
};

// create mock trip context
export function createMockTripContext(overrides = {}) {
  return {
    voteOnActivity: vi.fn(),
    votes: {},
    deleteComment: vi.fn(),
    members: [],
    addMember: vi.fn(),
    removeMember: vi.fn(),
    ...overrides,
  };
}

/**
 * Renders a component inside MemoryRouter
 */
export function renderWithRouter(ui) {
  return render(<MemoryRouter>{ui}</MemoryRouter>);
}

/**
 * Renders a component inside MemoryRouter + UserContext + TripContext
 */
export function renderWithAllProviders(
  ui,
  {
    userContext = defaultUserContext,
    tripContext = defaultTripContext,
    route = "/",
  } = {}
) {
  return render(
    <MemoryRouter initialEntries={[route]}>
      <UserContext.Provider value={userContext}>
        <TripContext.Provider value={tripContext}>{ui}</TripContext.Provider>
      </UserContext.Provider>
    </MemoryRouter>
  );
}

/**
 * Renders a component with just UserContext
 */
export function renderWithUserProvider(ui, { userContext } = {}) {
  return render(
    <UserContext.Provider value={userContext}>{ui}</UserContext.Provider>
  );
}
