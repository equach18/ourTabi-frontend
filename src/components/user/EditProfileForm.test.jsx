import { fireEvent, screen, waitFor } from "@testing-library/react";
import { vi } from "vitest";
import { Routes, Route } from "react-router-dom";
import EditProfileForm from "./EditProfileForm";
import { renderWithAllProviders } from "../../test-helpers/_testCommons";

const mockNavigate = vi.fn();

vi.mock("react-router-dom", async () => {
  const actual = await vi.importActual("react-router-dom");
  return {
    ...actual,
    useNavigate: () => mockNavigate,
  };
});

describe("EditProfileForm", () => {
  const testUser = {
    firstName: "first",
    lastName: "last",
    email: "first@example.com",
    profilePic: "https://example.com/test.jpg",
    bio: "test bio",
  };

  function renderForm() {
    return renderWithAllProviders(
      <Routes>
        <Route
          path="/profile/edit"
          element={
            <EditProfileForm
              updateUser={mockUpdateUser}
              currentUser={testUser}
            />
          }
        />
      </Routes>,
      { route: "/profile/edit" }
    );
  }
  const mockUpdateUser = vi.fn();

  beforeEach(() => {
    mockUpdateUser.mockReset();
    mockNavigate.mockReset();
  });

  it("matches snapshot", () => {
    const { asFragment } = renderForm();
    expect(asFragment()).toMatchSnapshot();
  });
  it("renders form fields", () => {
    renderForm();

    expect(screen.getByDisplayValue(testUser.firstName)).toBeInTheDocument();
    expect(screen.getByDisplayValue(testUser.lastName)).toBeInTheDocument();
    expect(screen.getByDisplayValue(testUser.email)).toBeInTheDocument();
    expect(screen.getByDisplayValue(testUser.profilePic)).toBeInTheDocument();
    expect(screen.getByDisplayValue(testUser.bio)).toBeInTheDocument();
  });

  it("shows error if no changes are made", async () => {
    renderForm();
    fireEvent.click(screen.getByRole("button", { name: /save changes/i }));

    expect(await screen.findByText(/no changes made/i)).toBeInTheDocument();
  });

  it("calls updateUser only when fields are changed", async () => {
    mockUpdateUser.mockResolvedValue({ success: true });

    renderForm();
    fireEvent.change(screen.getByPlaceholderText(/first name/i), {
      target: { value: "new name" },
    });

    fireEvent.click(screen.getByRole("button", { name: /save changes/i }));

    await waitFor(() => {
      expect(mockUpdateUser).toHaveBeenCalledWith({ firstName: "new name" });
      expect(mockNavigate).toHaveBeenCalledWith("/dashboard");
    });
  });

  it("shows error if updateUser fails", async () => {
    mockUpdateUser.mockResolvedValue({
      success: false,
      error: "Update failed",
    });

    renderForm();
    fireEvent.change(screen.getByPlaceholderText(/tell us about yourself/i), {
      target: { value: "New bio" },
    });
    fireEvent.click(screen.getByRole("button", { name: /save changes/i }));

    expect(await screen.findByText(/update failed/i)).toBeInTheDocument();
  });

  it("navigates to dashboard on cancel", () => {
    renderForm();
    fireEvent.click(screen.getByRole("button", { name: /cancel/i }));
    expect(mockNavigate).toHaveBeenCalledWith("/dashboard");
  });

  it("shows loading state while submitting", async () => {
    mockUpdateUser.mockImplementation(
      () => new Promise((res) => setTimeout(() => res({ success: true }), 100))
    );

    renderForm();
    fireEvent.change(screen.getByPlaceholderText(/last name/i), {
      target: { value: "new last name" },
    });
    fireEvent.click(screen.getByRole("button", { name: /save changes/i }));

    expect(screen.getByRole("button", { name: /saving/i })).toBeDisabled();
  });
});
