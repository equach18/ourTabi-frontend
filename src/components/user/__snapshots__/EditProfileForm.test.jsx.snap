// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditProfileForm > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="h-[calc(100vh-4rem)] flex items-center justify-center px-4 py-10"
  >
    <div
      class="w-full max-w-xl bg-white p-8 rounded-xl shadow-lg"
    >
      <h1
        class="text-3xl font-bold text-gray-600 mb-6"
      >
        Edit Profile
      </h1>
      <form
        class="space-y-4"
      >
        <input
          class="w-full py-2 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
          name="firstName"
          placeholder="First Name"
          type="text"
          value="first"
        />
        <input
          class="w-full py-2 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
          name="lastName"
          placeholder="Last Name"
          type="text"
          value="last"
        />
        <input
          class="w-full py-2 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
          name="email"
          placeholder="Email"
          type="email"
          value="first@example.com"
        />
        <input
          class="w-full py-2 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
          name="profilePic"
          placeholder="Profile Picture URL"
          type="url"
          value="https://example.com/test.jpg"
        />
        <textarea
          class="w-full py-2 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
          name="bio"
          placeholder="Tell us about yourself..."
          rows="4"
        >
          test bio
        </textarea>
        <div
          class="flex justify-end space-x-4 pt-2"
        >
          <button
            class="bg-zinc-400 text-white px-5 py-2 rounded-lg hover:bg-zinc-500 transition"
            type="button"
          >
            Cancel
          </button>
          <button
            class="bg-blue-500 text-white px-6 py-2 rounded-lg hover:bg-blue-600 transition false"
            type="submit"
          >
            Save Changes
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
